### PR TITLE
PrefixRange: canonicalize in constructor, cleanup

### LINKED ISSUE #728: Canonicalize PrefixRange in constructor
In the address, bits that don't matter for any length in the range should be set to 0s in the constructor.

For example, a PrefixRange made from 255.255.255.255/16:8-16 should be canonicalized to 255.255.0.0/16:8-16. I think 255.255.255.255/16:8-24 should be canonicalized to 255.255.255.0/16:8-24, but check with @dhalperi or @arifogel if you are not one of them.

### PR DESCRIPTION
**Pull request by dhalperi**
Stop extending Pair; the cleanliness in the constructor alone is worth it.

Fixes #728

---
This change is [Reviewable](https://reviewable.io/reviews/batfish/batfish/846)

### CODE REVIEW

File: projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/PrefixRange.java [hunk 1]
```diff
@@ -2,78 +2,50 @@
...
-public class PrefixRange extends Pair<Prefix, SubRange> {
+public final class PrefixRange implements Serializable, Comparable<PrefixRange> {
 
   /** */
   private static final long serialVersionUID = 1L;
 
+  public PrefixRange(Prefix prefix, SubRange lengthRange) {
+    // Canonicalize the prefix by dropping extra bits in the address that are longer than any
+    // relevant length.
+    int realPrefixLength = Math.min(prefix.getPrefixLength(), lengthRange.getEnd());
```
Comment (dhalperi) [inline projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/PrefixRange.java]: there is an open issue here as to whether this is precisely what's desired. @arifogel

**Comment by arifogel** [comment]
:lgtm:
One minor formatting comment.

---

Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/PrefixRange.java, line 19 at r1](https://reviewable.io:443/reviews/batfish/batfish/846#-L2XxMqS_Qv2wPaRgTiH-r1-19:-L2XxjoxBDAeLjyCg1Ee:b-jsn4d3) ([raw file](https://github.com/batfish/batfish/blob/8a73371b484fc379353aa134b8635fc549c1a601/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/PrefixRange.java#L19)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

there is an open issue here as to whether this is precisely what's desired. @arifogel 
</blockquote></details>

You would get the same result with `P/X:Y-Z` with `Z<X` as you would with `P/Z:Y-Z`. So to be canonical you should pick the smaller one. This looks correct.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/PrefixRange.java, line 83 at r1](https://reviewable.io:443/reviews/batfish/batfish/846#-L2XzEbcebmYeGd_3FVF:-L2XzEbdpCLhzppm1PO2:bk89cuf) ([raw file](https://github.com/batfish/batfish/blob/8a73371b484fc379353aa134b8635fc549c1a601/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/PrefixRange.java#L83)):*
> ```Java
>       return prefixCmp;
>     }
> ```

FYI I would omit the newline here, but your call.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/846#-:-L2Xzc87maRJ3uxNsKH2:baq10y4)*

**Comment by dhalperi** [comment]
Review status: 3 of 4 files reviewed at latest revision, 2 unresolved discussions.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/PrefixRange.java, line 83 at r1](https://reviewable.io:443/reviews/batfish/batfish/846#-L2XzEbcebmYeGd_3FVF:-L2Y2lXJYWrA05nF2jm5:b-896fix) ([raw file](https://github.com/batfish/batfish/blob/8a73371b484fc379353aa134b8635fc549c1a601/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/PrefixRange.java#L83)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

FYI I would omit the newline here, but your call.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/846)*

**Comment by dhalperi** [comment]
Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/846)*

**Pull request merged**